example/particle_leapfrog/particle_leapfrog.cpp

```diff
@@ -0,0 +1,455 @@
+
+  int max_active_index = swarm->GetMaxActiveIndex();
+
+  const auto &vx = swarm->Get<Real>("vx").Get();
+  const auto &vy = swarm->Get<Real>("vy").Get();
+  const auto &vz = swarm->Get<Real>("vz").Get();
+
+  // Assumes a grid with constant dx, dy, dz within a block
+  const Real &dx_i = pmb->coords.dx1f(0);
+  const Real &dx_j = pmb->coords.dx2f(0);
+  const Real &dx_k = pmb->coords.dx3f(0);
+  const Real &dx_push = std::min<Real>(dx_i, std::min<Real>(dx_j, dx_k));
+
+  auto swarm_d = swarm->GetDeviceContext();
+
+  Real min_dt;
+  pmb->par_reduce(
+      "particle_leapfrog:EstimateTimestep", 0, max_active_index,
+      KOKKOS_LAMBDA(const int n, Real &lmin_dt) {
+        if (swarm_d.IsActive(n)) {
+          Real v = sqrt(vx(n) * vx(n) + vy(n) * vy(n) + vz(n) * vz(n));
+          if (v != 0.0) {
+            lmin_dt = std::min(lmin_dt, dx_push / v);
+          }
+        }
+      },
+      Kokkos::Min<Real>(min_dt));
+
+  return cfl * min_dt;
+}
+
+} // namespace Particles
+
+// *************************************************//
+// define the application driver. in this case,    *//
+// that just means defining the MakeTaskList       *//
+// function.                                       *//
+// *************************************************//
+
+TaskStatus WriteParticleLog(MeshBlock *pmb) {
+  auto pkg = pmb->packages.Get("particles_package");
+
+  const auto write_particle_log = pkg->Param<bool>("write_particle_log");
+  if (!write_particle_log) {
+    return TaskStatus::complete;
+  }
+
+  auto swarm = pmb->swarm_data.Get()->Get("my particles");
+
+  const auto &id = swarm->Get<int>("id").Get().GetHostMirrorAndCopy();
+  const auto &x = swarm->Get<Real>("x").Get().GetHostMirrorAndCopy();
+  const auto &y = swarm->Get<Real>("y").Get().GetHostMirrorAndCopy();
+  const auto &z = swarm->Get<Real>("z").Get().GetHostMirrorAndCopy();
+  const auto &vx = swarm->Get<Real>("vx").Get().GetHostMirrorAndCopy();
+  const auto &vy = swarm->Get<Real>("vy").Get().GetHostMirrorAndCopy();
+  const auto &vz = swarm->Get<Real>("vz").Get().GetHostMirrorAndCopy();
+
+  const auto &is_active = swarm->GetMask().Get().GetHostMirrorAndCopy();
+  std::stringstream buffer;
+  for (auto n = 0; n < x.GetSize(); n++) {
+    if (!is_active(n)) {
+      continue;
+    }
+    buffer << Globals::my_rank << " , " << pmb->gid << " , " << id(n) << " , " << x(n)
+           << " , " << y(n) << " , " << z(n) << " , " << vx(n) << " , " << vy(n) << " , "
+           << vz(n) << std::endl;
+  }
+
+  std::cout << buffer.str();
+
+  return TaskStatus::complete;
+}
+
+// initial particle position: x,y,z,vx,vy,vz
+constexpr int num_test_particles = 4;
+const std::array<std::array<Real, 6>, num_test_particles> particles_ic = {{
+    {-0.1, 0.2, 0.3, 1.0, 0.0, 0.0},   // along x direction
+    {0.4, -0.1, 0.3, 0.0, 1.0, 0.0},  // along y direction
+    {-0.1, 0.3, 0.2, 0.0, 0.0, 1.0},  // along z direction
+    {0.12, 0.2, -0.3, 1.0, 1.0, 1.0}, // along diagonal
+}};
```

Comment:
This strikes me as very manual. Perhaps that's fine. But it might be worth thinking about if this could be automated.